File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:

        if len(s) < len(t):
            return ""

        res = ""
        letter_map = collections.defaultdict(int)
        needCount = len(t)

        for c in t:
            letter_map[c] = letter_map.get(c, 0) + 1

        left = 0
        for right in range(len(s)):
            if letter_map[s[right]] > 0:
                needCount -= 1

            letter_map[s[right]] -= 1

            while needCount == 0:
                if not res or len(res) > right-left+1:
                    res = s[left:right+1]

                letter_map[s[left]] += 1                

                if letter_map[s[left]] > 0:
                    needCount += 1
                    
                left += 1
        
        return res

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {

        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        int needCount = t.length();

        for (char c : t.toCharArray()){
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int startIndex = 0;
        int minLength = Integer.MAX_VALUE;

        int left = 0;
        for (int right=0; right < s.length(); right++){
            char c = s.charAt(right);
            if (map.containsKey(c) && map.get(c) > 0){
                needCount -= 1;
            }

            map.put(c, map.getOrDefault(c, 0) - 1);

            while (needCount == 0){
                if (minLength > right-left+1){
                    startIndex = left;
                    minLength = right-left+1;
                }

                map.put(s.charAt(left), map.get(s.charAt(left))+1);

                if (map.get(s.charAt(left)) > 0){
                    needCount += 1;
                }

                left += 1;
            }
        }

        if (minLength == Integer.MAX_VALUE)
            return "";

        return s.substring(startIndex, startIndex+minLength);
    }
}